File: practice/algorithms/bit_manipulation/twos_complement/twos_complement.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

vector<int> int2bin(int i_val)
{
    vector<int> res;
    while (i_val)
    {
        res.push_back(i_val & 1);
        i_val /= 2;
    }
    reverse(res.begin(), res.end());
    
    return res;
}

int bin2int(vector<int>::iterator i_begin,  vector<int>::iterator i_end)
{
    int res = 0;
    for (; i_begin != i_end; ++i_begin)
    {
        res = 2 * res + *i_begin;
    }
    return res;
}

unsigned long long num_bits(long long i_val)
{
    if (i_val == 0 || i_val == 1)
    {
        return i_val;
    }
    
    vector<int> bits = int2bin(i_val);
    const int n = bits.size();
    
    unsigned long long cnt = 0;
    for (int pos = 0; pos < n - 1; ++pos)
    {
        if (bits[pos] == 1)
        {
            cnt += (n - pos - 1) * pow(2ULL, n - pos - 2) + bin2int(bits.begin() + pos + 1, bits.end()) + 1;
        }
    }
    if (bits.back() == 1)
    {
        cnt++;
    }
    
    return cnt;
}

unsigned long long num_bits_neg(long long i_val)
{
    return 32ULL * abs(i_val) - num_bits(abs(i_val) - 1);
}

int main()
{
    int t, a, b;
    cin >> t;
    
    while (t--)
    {
        cin >> a >> b;
        if (a >= 1)
        {
            cout << (num_bits(b) - num_bits(a - 1)) << endl;
        }
        else if (a == 0)
        {
            cout << num_bits(b) << endl;
        }
        else if (b >= 0)
        {
            cout << (num_bits(b) + num_bits_neg(a)) << endl;
        }
        else if (b < -1)
        {
            cout << (num_bits_neg(a) - num_bits_neg(b + 1)) << endl;
        }
        else
        {
            cout << num_bits_neg(a) << endl;
        }
    }
    
    return 0;
}

File: practice/algorithms/bit_manipulation/twos_complement/twos_complement.py
#!/usr/bin/env python

def num_bits(i_val):
    if i_val == 0:
        return 0
    if i_val == 1:
        return 1
    
    bits = bin(i_val)[2:]
    n = len(bits)
    cnt = 0
    for POS in range(n - 1):
        if bits[POS] == '1':
            cnt += (n - POS - 1) * 2 ** (n - POS - 2) + int(bits[POS + 1:], 2) + 1
    if bits[-1] == '1':
        cnt += 1
    return cnt

def num_bits_neg(i_val):
    return 32 * abs(i_val) - num_bits(abs(i_val) - 1)

t = int(raw_input().strip())
for _ in range(t):
    A, B = map(int, raw_input().strip().split())
    if A >= 1:
        print num_bits(B) - num_bits(A - 1)
    elif A == 0:
        print num_bits(B)
    elif B >= 0:
        print num_bits(B) + num_bits_neg(A)
    elif B < -1:
        print num_bits_neg(A) - num_bits_neg(B + 1)
    else:
        print num_bits_neg(A)

